### PR TITLE
feat: add `--build-context` flag to task run

### PR DESCRIPTION
**Pull request by nbarrera**
this optional flag allows the user to provide the docker context path

if not provided, it fallbacks to the previous behaveior so this change
is transparently retro compatible, it takes the context from the
Dockerfile directory.

Addresses #3044

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/task_run.go [hunk 1]
```diff
@@ -909,6 +928,9 @@ Run a task with a command.
 			if cmd.Flags().Changed(dockerFileFlag) {
 				opts.isDockerfileSet = true
 			}
+			if cmd.Flags().Changed(dockerFileContextFlag) {
+				opts.isDockerfileContextSet = true
+			}
```
Comment (iamhopaul123) [inline internal/pkg/cli/task_run.go]: We probably don't need this as we dont' set the default value in https://github.com/aws/copilot-cli/pull/3046/files#diff-eec95b70c382971e4ec97c1db10feebe2b837deaba537fc9ec999e43ea62a9bbR946. We could simply check if it equals to "" to determine if this flag is set by users or not.

File: internal/pkg/cli/task_run.go [hunk 2]
```diff
@@ -751,9 +763,16 @@ func (o *runTaskOpts) buildAndPushImage() error {
...
+	if o.dockerfileContextPath == "" {
+		dockerContext = filepath.Dir(o.dockerfilePath)
+	} else {
+		dockerContext = o.dockerfileContextPath
+	}
+
 	if _, err := o.repository.BuildAndPush(dockerengine.New(exec.NewCmd()), &dockerengine.BuildArguments{
 		Dockerfile: o.dockerfilePath,
-		Context:    filepath.Dir(o.dockerfilePath),
+		Context:    dockerContext,
```
Comment (iamhopaul123) [inline internal/pkg/cli/task_run.go]: I wonder if we can set the default value for `o.dockerfileContextPath` in `newTaskRunOpts()` to have the default value to be
```go
if o.dockerfileContextPath == "" {
  o.dockerfileContextPath = filepath.Dir(o.dockerfilePath)
}
```
Then we can have `Context: o.dockerfileContextPath` here.

File: internal/pkg/cli/task_run.go [hunk 3]
```diff
@@ -310,12 +312,22 @@ func (o *runTaskOpts) Validate() error {
...
 		if _, err := o.fs.Stat(o.dockerfilePath); err != nil {
 			return err
 		}
 	}
 
+	if o.isDockerfileContextSet {
+		if _, err := o.fs.Stat(o.dockerfileContextPath); err != nil {
+			return err
+		}
+	}
```
Comment (iamhopaul123) [inline internal/pkg/cli/task_run.go]: After https://github.com/aws/copilot-cli/pull/3046/files#r751679516 we could simply have
```suggestion
        if o.image != "" {
        	if _, err := o.fs.Stat(filepath.Join(o.dockerfileContextPath, o.dockerfilePath)); err != nil {
		return err
		}
        }
```
to check if the dockerfile exists.

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/cli/task_run.go (hunk 1 above)
Comment (efekarakus) [reply internal/pkg/cli/task_run.go]: 👍 Done!

File: internal/pkg/cli/task_run.go (hunk 2 above)
Comment (efekarakus) [reply internal/pkg/cli/task_run.go]: I did this slightly differently but it's essentially the same if-statement

File: internal/pkg/cli/task_run.go (hunk 3 above)
Comment (efekarakus) [reply internal/pkg/cli/task_run.go]: It's possible that the Dockerfile and Context and in completely separate directories, that's why the stat checks are separate

**Review by efekarakus — APPROVED** [review]

**Comment by Lou1415926** [comment]
This is now released 🎉 in v1.14.0 https://github.com/aws/copilot-cli/releases/tag/v1.14.0 ! Thank you for the contribution!

**Pull request merged**